lib/matplotlib/collections.py

```diff
@@ -906,8 +906,11 @@ def update_scalarmappable(self):
         """
         if not self._set_mappable_flags():
             return
+        # We don't use the result of this call, but update the status for
+        # anyone watching.
+        self._check_update("array")
         # Allow possibility to call 'self.set_array(None)'.
-        if self._check_update("array") and self._A is not None:
+        if self._A is not None:
```

Comment:
Does this always return True? Otherwise, `self._mapped_colors` will not be calculated here, and has the default value `None`. In the following lines

https://github.com/matplotlib/matplotlib/blob/e6fd9012d823fde63db620412e1ce9c36d94eb63/lib/matplotlib/collections.py#L927-L934

the `_facecolors` and `_edgecolors`  may be set as `self._mapped_colors`, which is `None`. Then cause trouble in the evaluation in `draw`:
https://github.com/matplotlib/matplotlib/blob/e6fd9012d823fde63db620412e1ce9c36d94eb63/lib/matplotlib/collections.py#L379-L380

So I suggest to modify Lines 931 - 935 to
```python
      if self._face_is_mapped and self._mapped_colors is not None:
            self._facecolors = self._mapped_colors
      else:
            self._set_facecolor(self._original_facecolor)
      if self._edge_is_mapped and self._mapped_colors is not None:
            self._edgecolors = self._mapped_colors
```

Reply:
Actually, line 913 in the green could be replaced with
```
if self._face_is_mapped or self._edge_is_mapped:
```
because the check for `self._A is not None` is being done in the `_set_mappable_flags` helper immediately before this.
This change would make the logic clearer.

 There is no need in any case for the suggested modification of lines 931-935.

With the loss of the `check_update` call, the efficiency I was trying to achieve--recalculating the mapping only if something changed--is gone, so I think there is no longer a need to have the `self._mapped_colors` variable.  It can be local to the function, because it is always being recalculated if it is going to be used..